### ISSUE TITLE
Manually set location

### ISSUE THREAD

**Issue by smichel17** [LibreShift/red-moon#206]
Is it important that to be able to manually set your location?

I think the main advantage of using location is saving the effort of setting up the timer and keeping it in sync as the days grow longer and shorter. If manually setting your location takes more effort than you save, the feature is not worth offering; in order to be useful, it much be easy.

Figuring out your latitude and longitude is quite a bit of effort without using location services, so I don't think we should ask the user to do this. 

Setting your location from a map would work if the interface is good (ex: Twilight's map was very difficult to select a precise location with; I often had to try several times).

Alternatively, I wonder if it is possible to get a location based on the date and the sunrise/sunset times? If so, we could simply estimate a location based on the custom times the user has chosen and update the timer based on it. This seems like the easiest way, both from a user's point of view and effort to develop. However, it obviously depends on a big 'if'.

**Comment by nzmacalasdair** [comment]
So Redshift, a similar application, uses Latitude/Longditude measurements, which are pretty easy for a user to look up, with services like the following: https://www.latlong.net/,

Might be the easiest method to develop as well?

**Comment by decadent** [comment]
If I understand the question right, it's about having the option to manually specify the location, vs only using the system location?

If so, I'd like to add a datapoint: I have the system location services permanently disabled to avoid having the location snooped by Google. I know they still have it through other means, but with ever so slightly less precision.

As for the latitude and longitude, Wikipedia has it on cities' pages, and of course Google's search also can supply it.

**Comment by smichel17** [comment]
Red Moon does not care what your location is. We only ask for location because it can be used to calculate sunrise/sunset times, the data we actually care about. I am wondering if there is any other data we could ask the user for, that could be used to calculate sunrise/sunset times instead of location.

The reason I am interested is that this data might be for the user to produce/input than location. Prompting for Latitude/Longitude is certainly the easiest method to develop.

**Comment by Albirew** [comment]
An easier way would be to ask user to add city/country and to translate that internally into lat/lon, but for that, it would either need a database stored in app or to use a public database over the internets

**Comment by decadent** [comment]
Yeah, software like the desktop F.lux usually asks for the city, sometimes with auto-completion (which is much nicer than no autocompletion). The procedure is sorta worked-out on websites by now, but I'm not sure what kind of database is typically used—the db needs to be updated regularly and might be rather big to be embedded. One thing to be wary of is, there are plenty of places with identical names, so the region for each should be clarified in the auto-completion or in another way.

**Comment by Albirew** [comment]
Is there a need for regular updates and every villages of the world? I mean, big cities aren't moving, new cities are not created every months, and localisation data doesn't need to be really precise, using a city that is 50~100km² around you should be enough, no? If you're limiting database to that, it shouldn't be bigger than 1~2mb...

**Comment by decadent** [comment]
Well, for a commercial service it would be considered bad tone if a user can't find their town in the list—but it could fly for open-source software. Still, cities get renamed or change their population and databases fix their mistakes. People are notoriously dirty in their handling of real-world information and don't want to fit into computer-convenient constraints.

**Comment by Totorrr** [comment]
Hello, 

please consider adding an option to manually set the position using latitude an longitude values. I agree that a list of "big enough" cities location would be a cool plus. 

What I would also love, is the possibility of adding an offset to those sunset/sunrise times (like go red 30 or 60 minutes _after_ sunset), or to use twilight values: civil, nautical or astronomical dawn/dusk times. 

Like @decadent, I am reluctant to enable the location service when not really necessary, for battery, but mostly privacy concerns, and also because when I tweak Red Moon, it is usually at night, within a building, and I don't have a GPS signal (I disabled WiFi and GSM location services, because it sends sensitive info to Google servers and needs an Internet connection working). 

I am not in favor of "get a location based on the date and the sunrise/sunset times" as @smichel17 suggests, because: 

- it means you need to know the sunrise/sunset times for your location beforehand, as precisely as possible, 
- as we are close to the equinoxes (around March 21 and September 21), it would be impossible (or inaccurate at best) to guess the location of the device (the closer to the equinoxes, the less accuracy).

**Comment by smichel17** [comment]
See #281 —

- I think it is reasonable to add a text preference in which you can enter lat,lon. Not the ideal experience, but beats having to turn on gps when you don't want to.
    - I am willing to implement this, but no guarantees about when I will get to it, maybe in a month, maybe in a year or two. If you want it sooner, PRs welcome.
    - The bulk of the work is in just adding the android boilerplate for a preference, with the appropriate instructions. If someone implements just this part (but doing nothing on save), I will make it save to / load from the [correct place](https://github.com/LibreShift/red-moon/blob/master/app/src/main/java/com/jmstudios/redmoon/model/Config.kt).
- It is also reasonable to add a drop-down to select a city. I would suggest using the [IANA time zone](https://www.iana.org/time-zones) cities. However, adding both this *and* manual lat/lon would be excessive with the way the preferences are currently laid out. So, I would accept a PR adding this, but will not do it myself.

**Comment by ocouch** [comment]
I have an Android TV which has no location service. A manual location setting would be great.
